rundeck_action/action_types/rundeck.ts

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { curry } from 'lodash';
+import axios, { AxiosError, AxiosResponse } from 'axios';
+import { schema, TypeOf } from '@kbn/config-schema';
+import { nullableType } from '../../../x-pack/legacy/plugins/actions/server/builtin_action_types/lib/nullable';
+import { isOk, promiseResult, Result } from '../../../x-pack/legacy/plugins/actions/server/builtin_action_types/lib/result_type';
+import { ActionType, ActionTypeExecutorOptions, ActionTypeExecutorResult } from '../../../x-pack/legacy/plugins/actions/server/types';
+import { callPagerDutyApi, PagerDutyApiOptions, PagerDutyApiResult } from './lib/pagerduty'
+import { callSlackApi, SlackApiOptions, SlackApiResult } from './lib/slack'
+import { Logger } from '../logger'
+
+
+// config definition
+enum WebhookMethods {
+  POST = 'post',
+  PUT = 'put',
+}
+
+// Schema
+const headersSchema = schema.recordOf(schema.string(), schema.string());
+
+const configSchemaProps = {
+  rundeckBaseUrl: schema.uri(),
+  method: schema.oneOf([schema.literal(WebhookMethods.POST), schema.literal(WebhookMethods.PUT)], {
+    defaultValue: WebhookMethods.POST,
+  }),
+  rundeckApiVersion: schema.oneOf([schema.number()], { defaultValue: 24}),
+  jobId: schema.string(),
+  headers: nullableType(headersSchema),
+};
+
+type ActionTypeConfigType = TypeOf<typeof ConfigSchema>;
+const ConfigSchema = schema.object(configSchemaProps);
+
+type ActionTypeSecretsType = TypeOf<typeof SecretsSchema>;
+const SecretsSchema = schema.object({
+  rundeckApiToken: schema.string(),
+  pdApiKey: nullableType(schema.string()),
+  slackWebhookUrl: nullableType(schema.string()),
+});
+
+type ActionParamsType = TypeOf<typeof ParamsSchema>;
+const ParamsSchema = schema.object({
+  dedupKey: nullableType(schema.string()),
+  jobParams: schema.maybe(schema.object({
+    options: schema.maybe(schema.any())
+  })),
+});
+
+// action type definition
+export function getActionType({ logger }: { logger: Logger }): ActionType {
+  return {
+    id: '.rundeck',
+    name: 'rundeck',
+    validate: {
+      config: schema.object(configSchemaProps),
+      secrets: SecretsSchema,
+      params: ParamsSchema,
+    },
+    executor: curry(executor)({ logger }),
+  };
+}
+
+// action executor
+// * Need API Key for calling PD Rest APIs
+export async function executor(
```

Comment:
This function is super long and I totally get lost reading it. Would be nice to break this up a little.